keepalived-vip/controller.go

```diff
@@ -0,0 +1,253 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/intstr"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+)
+
+const (
+	reloadQPS     = 10.0
+	resyncPeriod  = 10 * time.Second
+	ipvsPublicVIP = "k8s.io/public-vip"
+)
+
+var (
+	// keyFunc for endpoints and services.
+	keyFunc = framework.DeletionHandlingMetaNamespaceKeyFunc
+
+	// Error used to indicate that a sync is deferred because the controller isn't ready yet
+	errDeferredSync = fmt.Errorf("deferring sync till endpoints controller has synced")
+)
+
+type service struct {
+	Ip   string
+	Port int
+}
+
+type vip struct {
+	Name     string
+	Ip       string
+	Port     int
+	Protocol string
+	Backends []service
+}
+
+// ipvsControllerController watches the kubernetes api and adds/removes
+// services from LVS throgh ipvsadmin.
+type ipvsControllerController struct {
+	queue             *workqueue.Type
+	client            *unversioned.Client
+	epController      *framework.Controller
+	svcController     *framework.Controller
+	svcLister         cache.StoreToServiceLister
+	epLister          cache.StoreToEndpointsLister
+	reloadRateLimiter util.RateLimiter
+	keepalived        *keepalived
+	reloadLock        *sync.Mutex
+}
+
+// getEndpoints returns a list of <endpoint ip>:<port> for a given service/target port combination.
+func (ipvsc *ipvsControllerController) getEndpoints(
+	s *api.Service, servicePort *api.ServicePort) (endpoints []service) {
+	ep, err := ipvsc.epLister.GetServiceEndpoints(s)
+	if err != nil {
+		return
+	}
+
```

Comment:
suggestion for future cleanup, we're duplicating the same loops in many service related controllers, we should isolate into a library interface if possible.